### DISCUSSION TITLE
Mechanism to Monkey Patch core files?

### DISCUSSION THREAD

**Discussion by Gavin-Holt** [orbitalquark/textadept#433] [Q&A] [answered]
Hi,

I have been playing with some alterations to the core lua files which define Textadept. e.g.

```lua
-- TextAdept\core\ui.lua:269
--   Change title format
local function set_title()
  local filename = buffer.filename or buffer._type or _L['Untitled']
  if buffer.filename then
    filename = select(2, pcall(string.iconv, filename, 'UTF-8', _CHARSET))
  end
  local basename = buffer.filename and filename:match('[^/\\]+$') or filename
  -- **
  ui.title = string.format('Textadept %s %s', filename:gsub([[\\]],[[\]]), buffer.modify and '*' or '-' )
  --**
  buffer.tab_label = basename..(buffer.modify and '*' or '')
end
```

and

```lua
-- TextAdept\core\file_io.lua:304
--    Use textredux
function io.open_recent_file()
  local utf8_list = {}
  for i = 1, #io.recent_files do
    utf8_list[#utf8_list + 1] = io.recent_files[i]:iconv('UTF-8', _CHARSET)
  end
  local results_list = textredux.core.list.new('Open Recent File')
    results_list.search_fuzzy = true
    results_list.column_styles = {
        textredux.core.style.string,
        textredux.core.style.string
    }
    results_list.headers = {
        'Files'
    }
    results_list.items = utf8_list
    results_list.on_selection = function(list, item)
        io.open_file(item)
        results_list:close()
    end
    results_list:show()
end
```
 and

```lua
-- TextAdept\modules\textredux\util\matcher.lua:95
--   Modifiy to only sort results if search_fuzzy=true
--    see https://github.com/rgieseke/textredux/issues/31
function M:match(search)
  if not search or #search == 0 then return self.candidates end
  local cache = self.cache
  if self.search_case_insensitive then search = search:lower() end
  local matches = cache.matches[search] or {}
  if #matches > 0 then return matches end
  local lines = cache.lines[string.sub(search, 1, -2)] or self.lines
  local matchers = self:_matchers_for_search(search)

  local matching_lines = {}
  for _, line in ipairs(lines) do
    local score = match_score(line.text, matchers)
    if score then
      matches[#matches + 1] = { index = line.index, score = score }
      matching_lines[#matching_lines + 1] = line
    end
  end
  cache.lines[search] = matching_lines

  -- **
  if self.search_fuzzy then
    table.sort(matches, function(a ,b) return a.score < b.score end)
  end
  -- **

  local matching_candidates = {}
  for _, match in ipairs(matches) do
    matching_candidates[#matching_candidates + 1] = self.candidates[match.index]
  end
  self.cache.matches[search] = matching_candidates
  return matching_candidates
end
```

I have previously done this all manually, not good when I download new releases!

I would like to make more alterations, but need a system to apply these to new releases.

The very reasonable advice to keep most module functions local, seems to prevent me from redefining them?

-   Is there any way to make this kind of change from my userdata\init.lua file?
-   Can anyone recommend an easy mechanism for monkey patching Textadept (for Windows 10 OS and a hobby programmer)?

Kind Regards Gavin Holt

PS I do not have, nor want; a compiler, GIT, VS Code, Electron anything! 
     I do have BusyBox v1.33.0 😄

**Comment by orbitalquark** [comment, marked as answer]
For `set_title()`, just take a look and see what event is calling it. For the latest nightly, it's 5 events: https://github.com/orbitalquark/textadept/blob/19cfa3ce0c7d2f981d2b8cbbee4f8e603be9b5db/core/ui.lua#L269-L274 and https://github.com/orbitalquark/textadept/blob/19cfa3ce0c7d2f981d2b8cbbee4f8e603be9b5db/core/ui.lua#L355-L361

All you have to do is connect to those events with your own `set_title()` in *~/.textadept/init.lua* and it'll be run after Textadept's version.

For `io.open_recent_file()`, simply define your own `io.open_recent_file()` and change the menu option in *~/.textadept/init.lua*:

```
function io.open_recent_file()
  ...
end
textadept.menu.menubar['File/Open Recent...'][2] = io.open_recent_file
```

Note that the latest nightly allows for this menu syntax. Previous versions of Textadept need the `_L` table for getting the localized version. If you're using English, then it probably won't matter. You may also have to edit any other outside references to `io.open_recent_file` at the file-level. (Anything inside functions will use your version.)

I don't know about your `M:match()` example because I'm not familiar with TextRedux. I suspect you might be able to redefine it using something like `require('textredux.util.matcher').match = function(self, search) ... end` in your *~/.textadept/init.lua*.
